### DISCUSSION TITLE
Timeline/time-thread to display important notes/events

### DISCUSSION THREAD

**Discussion by j-steinbach** [srid/emanote#141] [Ideas]
My personal notes are a mix of 

- daily journal and diary
- personal wiki/knowledge base (PKB/W)
- todo-list
- Zettelkasten (ZK).

Generally, the PKW and ZK contains *meaningful*, time-independent notes, whereas the journal and diary contains mostly time/date-specific and repetitive non-information pertaining to events. Sometimes there are events that are very important/noteworthy though.

Ideally I could annotate/highlight these important events and add them to a timeline/time-thread. This timeline would only feature marked notes (maybe tagged "important" or "noteworthy") and display them in a chronological order (based on their associated  date).

This would allow taking multiple daily notes/journal entries ("dinner with in-laws", "got a parking ticket", "aunt mays birthday", "met josh on the street") and curate only the important ones to appear on the timeline. [Remember, *meaningful* notes.]

Furthermore, this could also be used to "pin" events that will appear in the future. Or display repeating events (show the PKW entry for a person that has a birthday coming up).

**Comment by srid** [comment]
There is something called 'timeline query', as demonstrated by https://www.srid.ca/blog

Source: https://raw.githubusercontent.com/srid/www.srid.ca/master/Blog.md

(Obsidian doesn't support search flags, so I ended up overloading the markdown class specifier syntax -- `{.timeline}`).

Do you have a concrete proposal of how this can be improved?

**Comment by j-steinbach** [comment]
![emanote-timeline](https://user-images.githubusercontent.com/69524139/134378215-2f31b3d9-dc97-48e8-a48e-a222469b9c67.png)

I have some ideas how this might/should look, but I am not sure how to implement it.

In my opinion there is a difference between a simple query and a curated timeline. I *choose* what to put on the timeline, whereas the query just stupidly displays whatever I ask it for. (Obviously, a query can be fine-tuned to get me what I want.) 

Maybe it is helpful to think of the timeline as a specific, curated and enhanced query + eye candy.

**Reply by srid** [reply]
Interesting.

It looks like your idea is a special case of https://github.com/srid/emanote/discussions/106 

By implementing the obsidian-dataview, we should be able to support arbitrary queries / views such as the above. Definitely take a look at [their design](https://github.com/blacksmithgu/obsidian-dataview#obsidian-dataview), it looks interesting in a more general sense.

**Reply by srid** [reply]
In particular, I'm envisioning that the user should be able to write their own HTML template that renders the results of a query. The timeline query is implemented in this way already: https://github.com/srid/emanote/blob/master/default/templates/filters/query-timeline.tpl 

If you add a `templates/filters/query-timeline.tpl` in your notebook, you can definite how to render the timeline query. 

A more general idea is to support multiple query times, and for each query type allow the user (or default layer) to define how to render the results in HTML.

**Reply by j-steinbach** [reply]
Hmm, I guess you can understand it that way, and the custom HTML templates are a great idea.

----

But I am not fully convinced that these things aren't different -- I guess I am just having a hard time putting my "minds-eyes view" into text/pictures. I will try to describe my thoughts so that I can come revisit them later and not forget about it:

- There is a time bar/line. It is empty
- I can pin specific notes to the bar. Like [this](https://forbo.blob.core.windows.net/forboimages/20502/1180x664_BulletinBoard_2106_TIMELINE.jpg)?
- Notes should cluster around their respective dates. A single date can have multiple notes. 
- The notes/bar should have rules to allow highlighting/marking them.

Maybe the difference lies in the act of "pinning"? It should be possible to mix events (associated with a date) and articles from my PKB/ZK.

**Comment by srid** [comment]
A bit tangential: if one knows Haskell, this sort of "custom" views can be achieved on top of https://github.com/srid/ema/pull/81

I have a proof-of-concept in my private notebook.[^1] Soon I'll publish a blog post explaining it.

[^1]: I use level-2 headings with dates in them, across multiple notes. This is how I write "daily" notes - but in the context of that note. So I might have a note to track food sensitivities, and each level 2 heading would be a day on which I changed my diet in a certain way with a paragraph detailing it. Over time I have a 'timeline' of food changes. And I might have another note on the work timeline, etc.

**Reply by srid** [reply]
Basically, one writes a Haskell function that takes the Emanote model (all the data loaded from disk, the Pandoc AST, metadata, etc.) and returns the HTML for some page.
